### PR TITLE
emacs: Prevent Macports tree-sitter grammar libs from shadowing

### PR DESCRIPTION
**Pull request by wyuenho**
#### Description

~~The reason for this change is that the canonical way to use the new tree-sitter functionality in Emacs 29 is to either configure Emacs yourself to have it compile the libraries for you automatically or use a package like [treesit-auto](https://melpa.org/#/treesit-auto). Emacs upstream frequently assumes whatever the HEAD was in the tree-sitter grammar repo when the Elisp code was written and pays little attention to the differences between HEAD and the last released version, which frequently falls far behind HEAD by years and thus miss out on lots of the latest changes in the languages. There's also an ever increasing number of these language libraries but no reasonable Emacs users will ever require all of them. Let's make these a do-it-yourself part of Emacs as the users are so accustomed to.~~

The `treesit-extra-load-path` defined in emacs-app and emacs-app-devel are shadowing user-compiled tree-sitter grammar libraries in `user-emacs-directory`. This PR is the fix.

In addition, this PR also reverts 2416784.

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 14.5
Xcode 15.4

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by amake** [comment]
> The reason for this change is that the canonical way to use the new tree-sitter functionality in Emacs 28 is to either configure Emacs yourself to have it compile the libraries for you automatically or use a package like [treesit-auto](https://melpa.org/#/treesit-auto).

I will admit that I am not perfectly tuned in to everything on emacs-devel, but I do peek in occasionally, and all of this sounds backwards to me:

I thought external packages were the *old* way to use tree-sitter, and then tree-sitter support was brought into core. Then, the intention was for package managers (us) to handle the necessary configuration on behalf of users.

> There's also an ever increasing number of these language libraries but no reasonable Emacs users will ever require all of them. 

I don't think this is a problem because they are all tiny. I would rather provide all of them than make a user figure out how to do this on their own.

> Let's make these a do-it-yourself part of Emacs as the users are so accustomed to.

I feel like the balance points the other way:

- The cost to us providing everything out of the box (maintenance of the tree-sitter ports, keeping track of which are used by built-in `*-ts-mode`s) is small
- The cost to the user (extraneous packages installed) is small
- The benefit to the user is potentially large
- If we were to *stop* providing this, the cost to the user would be large (my sense is that users find this notoriously difficult to set up)

> Emacs upstream frequently assumes whatever the HEAD was in the tree-sitter grammar repo when the Elisp code was written and pays little attention to the differences between HEAD and the last released version, which frequently falls far behind HEAD by years and thus miss out on lots of the latest changes in the languages.

Are you saying that the Emacs `*-ts-mode`s are only compatible with some past snapshot of tree-sitter-* libraries? That doesn't seem to match my understanding of how tree-sitter libraries work. I thought that by definition they presented a stable API.

The only compatibility problem I am aware of is that recently tree-sitter broke ABI(?) compatibility without bumping their installed version number, or something like that. That's definitely something that package managers should be handling, not end users.

**Comment by amake** [comment]
Jumping out ahead in case everyone disagrees with me: a compromise I would be OK with:

- Remove the depends_run libraries
- Keep the `treesit-extra-load-path` preset (this is very important)
- Add a note instructing the user about installing tree-sitter runtime libraries

**Comment by wyuenho** [comment]
> I thought external packages were the old way to use tree-sitter, and then tree-sitter support was brought into core. Then, the intention was for package managers (us) to handle the necessary configuration on behalf of users.

> Are you saying that the Emacs *-ts-modes are only compatible with some past snapshot of tree-sitter-* libraries? That doesn't seem to match my understanding of how tree-sitter libraries work. I thought that by definition they presented a stable API.

That's not true. Tree-sitter support is experimental at best at the time of Emacs 29's release. You can browse the [commits and their bug tickets](https://github.com/emacs-mirror/emacs/commits/emacs-29) on the emacs-29 branch and see how many bug fixes to those *-ts-modes were applied because of the grammar queries don't work on released versions of tree-sitter language libraries or were inflexible enough that when new versions of the libraries were installed, the queries would break. I have personally filed at least half a dozen of these bugs since Emacs 29 was released.

Also, if you read the [release notes](https://github.com/emacs-mirror/emacs/blob/emacs-29/etc/NEWS), there are subtle references to this problem.

> You only need to install language grammar libraries required by the
Emacs modes you will use, as Emacs loads these libraries only when the
corresponding mode is turned on in some buffer for the first time in
an Emacs session.

> We generally recommend to use the latest versions of grammar libraries
available from their sites, as these libraries are in constant
development and occasionally add features and fix important bugs to
follow the advances in the programming languages they support.

Given that Macport's policy is to not provide unreleased versions of packages unless specified, the alternative is to provide -devel versions for all of the tree-sitter language libraries and have Emacs depend on them. A second alternative is to provide a separate optional variant where only the tree-sitter language libraries are depended on by Emacs, whereas the tree-sitter core remains part of the default variant.

> I feel like the balance points the other way:

The cost to us providing everything out of the box (maintenance of the tree-sitter ports, keeping track of which are used by built-in *-ts-modes) is small
The cost to the user (extraneous packages installed) is small
The benefit to the user is potentially large
If we were to stop providing this, the cost to the user would be large (my sense is that users find this notoriously difficult to set up)

I agree and I have made similar arguments to emacs-devel years ago, but that I was reprimanded and reminded of the fact that I was using Emacs and it's not suppose to be user-friendly to the 99%, but only customizable to the 100%.

> Keep the treesit-extra-load-path preset (this is very important)

No. Users have different preferences as to where to put their libraries, especially those that tend to be self-compiled. This has tripped me up when I was debugging why some -ts-modes were able to launch but queries were incorrect, even when I didn't install any tree-sitter library for that language. This implicit supposed convenience has made debugging even harder.

This is from the release notes:

> Then place the shared library
you've built in the same directory where you keep the other shared
libraries used by Emacs, or in the "tree-sitter" subdirectory of your
'user-emacs-directory', or in a directory mentioned in the variable
'treesit-extra-load-path'.

> Add a note instructing the user about installing tree-sitter runtime libraries from MacPorts

Good idea.

**Comment by amake** [comment]
I'll just say that I disagree: this PR makes most people's lives worse to the benefit of very few.

But I will abide by the decision of the port maintainers.

**Comment by wyuenho** [comment]
> Especially in that I expect MacPorts-installed Emacs to pick up MacPorts-installed tree-sitter-* libraries out of the box.

There are 2 reasons I think this line of thinking is flawed.

1) There are many many Emacs packages already require installation of external tools or libraries. Any Emacs users with > 1 week of experience should be used to this. Installing extra tree-sitter grammar libraries will already be automatically picked up by Emacs without modifying `treesit-extra-load-path`.
2) Nobody requires all 23 grammar libraries.

I have thought about just requiring the grammar libs that have corresponding -ts-modes built into Emacs, but due to the reasons stated [here](https://github.com/macports/macports-ports/pull/24241#issuecomment-2143209259), I'm convinced that it's futile at the moment. We may be able to do this in a future where the quality of the built-in -ts-modes has improved.

> We are not emacs-devel; we are empowered to provide the Emacs experience we think our users want, even if that's not exactly what emacs-devel wants.

You are not making any Emacs users' lives easier unless you can guarantee every grammar lib version available on Macports will work flawlessly with any Macports provide versions of Emacs.

**Comment by wyuenho** [comment]
The reason you thought `treesit-extra-load-path` was needed was because a `dlopen` call inside the macOS app bundle doesn't automatically search into Macport's prefix, so the tree-sitter grammar libs couldn't be picked up. You thought you could work around it like how the `site-start.el` modifies `load-path`. The problem is, the `load-path` modified works as long as the users hasn't installed any Emacs packages from Macports, their own package installed via Melpa will not be shallowed. However, `treesit-extra-load-path` is first looked up by Emacs and the search stops as soon as a language library is found. Since the variable takes a list, and there's no easy way to know what the user has defined `user-emacs-directory` to be at compile time, you've basically overridden the user's intention to have Emacs load up the grammar libraries compiled in `user-emacs-directory` by setting the variable with the highest priority.

I've just pushed a fix for this by adding another -rpath ld flag to emacs-app and emacs-app-devel to point to Macport's prefix. This has an additional advantage to not disrupt existing emacs-apps installations when the users upgrade to the new revision and everything will continue to work. It's just in the future, the users are suppose to deal with installing tree-sitter grammar libs manually.

File: editors/emacs/files/site-start-app.el
Comment (wyuenho) [inline editors/emacs/files/site-start-app.el:14]: In case anybody wonders, I've mistakenly remove this in 2416784c00f9a292768e12cc2233cb252290aa15. This is still very much needed for Emacs 29.

**Comment by pmetzger** [comment]
So, I have to admit I'm not entirely happy about the idea of making the user deal with the tree-sitter stuff on their own; it makes the port a lot less useful, at least to me?

**Comment by pmetzger** [comment]
(I can probably be convinced but at the moment this seems like a bit of a lose? It means that every time Emacs updates for me I'm going to have to notice and perform an action by hand that I'm not familiar with.)

**Comment by wyuenho** [comment]
> So, I have to admit I'm not entirely happy about the idea of making the user deal with the tree-sitter stuff on their own; it makes the port a lot less useful, at least to me?

Nobody is happy with how tree-sitter got integrated into Emacs 29, but [this is even worse](https://github.com/macports/macports-ports/pull/24241#issuecomment-2143537356).

**Comment by wyuenho** [comment]
> (I can probably be convinced but at the moment this seems like a bit of a lose? It means that every time Emacs updates for me I'm going to have to notice and perform an action by hand that I'm not familiar with.)

As stated in the description, you can just use [treesit-auto](https://melpa.org/#/treesit-auto). In case you don't want to, you really only have to call `treesit-install-language-grammar` with perhaps some minor tweaks to `treesit-language-source-alist` every couple of months. The point of this PR is to make the above options work without the users digging into why their self-compiled grammar libs got shadowed.

**Comment by pmetzger** [comment]
"You can just use" means there's something I have to do by hand that I would hope the package install would do for me?

**Comment by amake** [comment]
> I've just pushed a fix for this by adding another `-rpath` `ld` flag to `emacs-app` and `emacs-app-devel` to point to Macports prefix.

> Installing an extra grammar is just one `port install tree-sitter-[lang]` away.

So setting `treesit-extra-load-path` to point to the MacPorts prefix is no longer needed **and** MacPorts-installed tree-sitter-* libraries will be found automatically?

If that's the case then my main objections are addressed.

**Comment by pmetzger** [comment]
So after this is merged, how do I get a functioning emacs including tree sitter?

**Comment by wyuenho** [comment]
> Is there a way we could have this (apparently better) way of doing things automated so it didn't need to be thought about?

I *could* put the universe of tree-sitter grammar lib deps back now that my own self-compiled grammar libs won't be shadowed, but I'm semi-repulsed by this all-or-nothing approach, but the alternative is to define 20+ and counting many variants and the users only pick those they want so Macports can update them automatically during `port upgrade outdated`. I'm also semi-repulsed by this.

> So after this is merged, how do I get a functioning emacs including tree sitter?

They will continue to work. You can mark the grammar libs you are interested with `port setrequested tree-sitter-[lang]` so they don't get deleted when you do a `port reclaim`.

Or install [treesit-auto](https://github.com/renzmann/treesit-auto/tree/016bd286a1ba4628f833a626f8b9d497882ecdf3) from Melpa, or [read this article](https://www.masteringemacs.org/article/how-to-get-started-tree-sitter) and get down and dirty.

**Comment by pmetzger** [comment]
Why is it repulsive to just install the tree sitter libraries if people want them?

**Comment by wyuenho** [comment]
Who are these people who require 20+ grammar libs? Do they exist?

**Comment by pmetzger** [comment]
No, but the grammar libraries are very tiny compared to the disk space on my computer, or even to the total size of emacs, and it's simpler just to have them all pre-built than to figure out what I want.

**Comment by wyuenho** [comment]
But you won't have them all, there are dozens more of these grammar libs not required by the Emacs port that dozens other new Melpa packages require. With treesit-auto tho, you'll almost have them all.

**Comment by amake** [comment]
> the alternative is to define 20+ and counting many variants and the users only pick those they want

My problem with any variant-based solution is that unless they are default variants, customizing the selection of variants prevents you from using MacPorts's prebuilt binaries, and Emacs is heavy enough to make that very annoying.

The theory behind the current selection of deps is to include everything that Emacs has a built-in *-ts-mode for. One way to split this baby is to, on top of the changes in this PR, add a new "treesitter_libs" variant that just adds the runtime deps. Due to the above I would want it to be a default variant, though. Then users who really don't want to spend the 3.2M* to have all the libs can get what they want, though at the cost of building from source.

\* `du -shc /opt/local/lib/libtree-sitter-*` on my machine, with I think *all* tree-sitter-* ports installed, which is more than Emacs pulls in.

**Comment by amake** [comment]
> Who are these people who require 20+ grammar libs? Do they exist?

Who are the people who require all of the various Elisp packages bundled with Emacs?

I think it's odd to have functionality in Emacs that doesn't work unless you go and install more libraries from the same place you got Emacs. That is very much something I expect my package manager to give me out of the box.

**Comment by amake** [comment]
> The reason you thought `treesit-extra-load-path` was needed was because...

For what it's worth, this paragraph does correctly describe how I arrived at the current shape of the treesitter variant. If nothing else, we should indeed merge the correction to how the treesitter libraries are resolved.

**Comment by wyuenho** [comment]
Ok I've put the grammar libs required by the built-in *-ts-modes back and changed the commit message and PR description. This is now purely a bugfix PR.

File: editors/emacs/Portfile
Comment (amake) [inline editors/emacs/Portfile:346]:
```suggestion
To install tree-sitter grammar libraries not required by built-in *-ts-modes,
```

File: editors/emacs/Portfile
Comment (amake) [inline editors/emacs/Portfile:111]: I think emacs{,-app}-devel also needs a revision bump.

File: editors/emacs/Portfile
Comment (wyuenho) [reply editors/emacs/Portfile]: fixed

**Review by amake — APPROVED** [review]

**Comment by pmetzger** [comment]
@wyuenho Thank you. I know it's sometimes unpleasant having these sorts of discussions; we all are always sure that we're right and the other person is wrong, and I have lost many of them myself and not felt great about it. However, these discussions are valuable, and I wanted to tell you that I appreciate how good humored you have been about this.

**Comment by pmetzger** [comment]
@drkp Any thoughts before we commit this?

**Pull request merged**
